enclave/core/sgx/calls.c

```diff
@@ -775,10 +830,71 @@ oe_result_t oe_call_host_function_internal(
     /* Check the result */
     OE_CHECK(args->result);
 
+    return_args_ptr = (oe_call_function_return_args_t*)output_buffer;
+    /* Copy the marshaling struct from the host memory to avoid TOCTOU issues.
+     */
+    memcpy(
+        &return_args, return_args_ptr, sizeof(oe_call_function_return_args_t));
+
+    if (return_args.result == OE_OK)
+    {
+        /*
+         * Error out the case if the deepcopy_out_buffer is NULL but the
+         * deepcopy_out_buffer_size is not zero.
+         * Note that this could only happen if the oeedger8r-generated routine
+         * is modified.
+         */
+        if (!return_args.deepcopy_out_buffer &&
+            return_args.deepcopy_out_buffer_size)
+            OE_RAISE(OE_UNEXPECTED);
+
+        /*
+         * Nonzero deepcopy_out_buffer and deepcopy_out_buffer_size fields
+         * indicate that there is deep-copied content that needs to be
+         * transmitted from the host.
+         */
+        if (return_args.deepcopy_out_buffer &&
+            return_args.deepcopy_out_buffer_size)
+        {
+            /*
+             * Ensuring that the content lies on host memory.
+             * Note that this could only happen if the oeedger8r-generated
+             * routine is modified.
+             */
+            if (!oe_is_outside_enclave(
+                    return_args.deepcopy_out_buffer,
+                    return_args.deepcopy_out_buffer_size))
+                OE_RAISE(OE_UNEXPECTED);
+
+            void* enclave_buffer =
+                oe_malloc(return_args.deepcopy_out_buffer_size);
+            /* Copy the deep-copied content to enclave memory. */
+            memcpy(
+                enclave_buffer,
+                return_args.deepcopy_out_buffer,
+                return_args.deepcopy_out_buffer_size);
+            /* Release the memory on host heap. */
+            oe_host_free(return_args.deepcopy_out_buffer);
+            /*
+             * Update the deepcopy_out_buffer field.
+             * Note that the field is still on host memory. Currently, the
+             * oeedger8r-generated code will perform additional check that
+             * ensures the buffer stays within the enclave memory.
+             */
+            return_args_ptr->deepcopy_out_buffer = enclave_buffer;
+        }
+    }
+
     *output_bytes_written = args->output_bytes_written;
     result = OE_OK;
 
 done:
+    if (result != OE_OK)
+    {
+        return_args_ptr->result = result;
```

Comment:
Do we ensure that return_args_ptr lies out side the enclave?

Reply:
We didn't validate the `return_args_ptr`, or `output_buffer` here because the buffer is currently allocated via `oe_allocate_ocall_buffer` and then is passed to here.
```
    _buffer = (uint8_t*)oe_allocate_ocall_buffer(_total_buffer_size);
    _input_buffer = _buffer;
    _output_buffer = _buffer + _input_buffer_size;
    if (_buffer == NULL)
    {
        _result = OE_OUT_OF_MEMORY;
        goto done;
    }
```

Given that is is all enclave-side code, the host cannot modify the value of `output_buffer`.